urla958522fc1_spirosikmd_inmdsy_08_tgz-pJ8-TesterAppJ8/src/nl/rug/peerbox/logic/Peerbox.java

```diff
@@ -0,0 +1,207 @@
+import nl.rug.peerbox.middleware.MessageListener;
+import nl.rug.peerbox.middleware.Multicast;
+import nl.rug.peerbox.middleware.PrettyPrinter;
+import nl.rug.peerbox.middleware.ReliableMulticast;
+
+import org.apache.log4j.Logger;
+
+public class Peerbox implements MessageListener, Context {
+
+	private static final String PEERBOX_PROPERTIES_FILE = "peerbox.properties";
+	private static final String DEFAULT_PROPERTIES_FILE = "default.properties";
+	private final static Logger logger = Logger.getLogger(Peerbox.class);
+
+	private Multicast group;
+	private String path;
+	private String datafile;
+	private ExecutorService pool = Executors.newFixedThreadPool(5);
+	private Peer peer;
+	private VirtualFileSystem fs;
+	private PeerManager peerManager;
+
+	private static class Holder {
+		private static final Context INSTANCE = Peerbox.createInstance();
+	}
+
+	public static synchronized Context getInstance() {
+		return Holder.INSTANCE;
+	}
+
+	private Peerbox() {
+		Properties defaultProperties = new Properties();
+		createDefaults(defaultProperties);
+
+		Properties properties = new Properties(defaultProperties);
+		if (new File(PEERBOX_PROPERTIES_FILE).exists()) {
+			try (FileInputStream in = new FileInputStream(
+					PEERBOX_PROPERTIES_FILE)) {
+				properties.load(in);
+			} catch (IOException e) {
+				logger.error(e);
+			}
+		}
+
+		path = properties.getProperty(Property.PATH);
+		datafile = properties.getProperty(Property.DATAFILE_NAME);
+		String address = properties.getProperty(Property.MULTICAST_ADDRESS);
+		int port = Integer.parseInt(properties
+				.getProperty(Property.MULTICAST_PORT));
+		group = ReliableMulticast.createPeer(address, port);
+		peerManager = new PeerManager(group);
+		byte[] ip = getLocalAddress();
+		int serverPort = Integer.parseInt(properties
+				.getProperty(Property.SERVER_PORT));
+		String name = properties.getProperty(Property.NAME);
+		peer = Peer.createPeer(ip, serverPort, name);
+	}
+
+	private static Peerbox createInstance() {
+		Peerbox peerbox = new Peerbox();
+		peerbox.group.addMessageListener(peerbox);
+		peerbox.fs = VirtualFileSystem.initVirtualFileSystem(peerbox);
+		Thread server = new Thread(new FileServer(peerbox));
+		server.setDaemon(true);
+		server.start();
+		
+		PrettyPrinter.registerPrinter(new Message.MessagePrinter());
+		
+		return peerbox;
+	}
+
+	private static byte[] getLocalAddress() {
+		byte[] ip = new byte[] {};
+		try {
+			ip = InetAddress.getLocalHost().getAddress();
+		} catch (UnknownHostException e) {
+			logger.error(e);
+		}
+		return ip;
+	}
+
+	@Override
+	public void join() {
+		Message message = new Message();
+		message.put(Key.Command, "JOIN");
+		message.put(Key.Peer, getLocalPeer());
+		group.announce(message.serialize());
+	}
+
+
+	@Override
+	public void leave() {
+		Message leave = new Message();
+		leave.put(Key.Command, Command.ByeBye);
+		leave.put(Key.Peer, getLocalPeer());
+		group.announce(leave.serialize());
+		pool.shutdownNow();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Call `shutdown()` to reject any incoming tasks and `awaitTermination()` to wait for existing tasks to complete or until a timeout is reached. To stop the execution and processing of tasks, if necessary, call `shutdownNow()`. Check the return value of `awaitTermination()` and handle when the method times out while waiting for other threads to stop executing. See the following diffs on GitHub that make this code change: [diff1](https://github.com/glowroot/glowroot/commit/81de5a3699c59e0cf98d485c380c1034e6cd28a5#diff-c608d2c40790b0e0e5228e6647f69729caa756bbe9127169139406314b80897eL143), [diff2](https://github.com/apache/directory-kerby/commit/3ed0e7cd3b8a0c3ce29293153e4e72f422f0637f#diff-eb42f072849f520563ec2a686789a836195f18075cc222ef355c6ea710b0a360L84)

Learn more: [ExecutorService in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html)